libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.cpp

```diff
@@ -727,11 +727,13 @@ int HTTPClient::writeToStream(Stream * stream)
     if(_transferEncoding == HTTPC_TE_IDENTITY) {
         if(len > 0) {
             ret = writeToStreamDataBlock(stream, len);
+        } else if(len == -1 && _client->available() > 0) {
+            ret = writeToStreamDataBlock(stream, _client->available());
+        }
```

Comment:
Could you please test this ?
```
        if (len > 0 || len == -1) {
            ret = writeToStreamDataBlock(stream, len);
        } // no else
```
`writeToStreamDataBlock()` is supposed to correctly handle `-1`.
`available()>0` does not garantee that every incoming data have arrived.
When `-1` is given it's either `HTTP/1.0` or `Connection: Close`. In both case, every data must be swallowed until connection is closed by peer.
I'm not sure `writeToStreamDataBlock()` handles this without a timeout though.

This will fix #7692

Reply:
I see your point about possibly not consuming all data in the response body.  The response body coming from the device I am utilizing (a Denon AV receiver) isn't large, so it's always working (and did not work prior to the change I suggested).  But you're right, it should work in all circumstances.

I tested your suggested change, and it alone is not enough to solve the original issue.  The problem appears to have something to do with the code in `writeToStreamDataBlock()`, which sets `readBytes` to `len`, and then passes that to `_client->readBytes()`, which presumably isn't expecting a negative length.

```
    // read all data from server
    while(connected() && (len > 0 || len == -1))
    {
        int readBytes = len;

        // not read more the buffer can handle
        if(readBytes > buff_size) {
            readBytes = buff_size;
        }

        // read data
        int bytesRead = _client->readBytes(buff, readBytes);

```
Thus, my original suggestion solved the problem, at least for response bodies that fit and are fully in buffer by the time the call to `writeToStreamDataBlock()` occurs in `writeToStream()`.  To solve it for all circumstances, perhaps a combination of your suggestion in `writeToStream()` and a change in the above `while` loop that utilizes `_client->available()` like my original suggestion, but does it repeatedly until there's no more to read, is appropriate?

Reply:
I too have been confused with `bytesRead` and `readBytes`.
To summarize the function:

- `int size` goes to `int len`
- `int readBytes` takes `len`

loop:
- `int bytesRead` takes `read(readBytes)` (possible timeout when input stream is over => end of transmission)
- `write(bytesRead)`
- `len -= bytesRead`

The `if (len<0)` was introduced in #6476 between releases 2.7.1 and 2.7.2.

> it should work in all circumstances.

:+1:

> I tested your suggested change, and it alone is not enough to solve the original issue. 

Can you be more specific, does it work after a timeout or doesn't work at all ?


I am also interested to know if your issue is solved by #6979 (also included in [alpha release v0.0.2](https://d-a-v.github.io/esp8266/Arduino)) which is a complete rewrite of block transfer functions, that are currently duplicated all over the place and not necessarily efficient.

Reply:
> perhaps a combination of your suggestion in writeToStream() and a change in the above while loop that utilizes `_client->available()` like my original suggestion, but does it repeatedly until there's no more to read, is appropriate?

When `_client` is a tcp connection (`WiFiClient`), `::available()` can be 0 at the time of testing and >0 a bit later.
So instead, the connection status needs to be checked (ie: has peer closed?) specifically when `len` is infinite (-1). That could be added in the current code.

Reply:
Thanks for the explanations.  After re-reading everything very carefully, and in an effort to answer your inquiries and upon further testing, I need to apologize and retract what I said above about your original suggestion alone not being enough to solve problem.  It does so, perfectly.

The bottom line is that, as I think you suspected, `writeToStreamDataBlock()` handles this with a timeout.  That makes sense.  If you don't have a Content-Length, you have to decide at some point when you're going to consider what you've got thus far to be _the_ response.  Turns out I hadn't proactively managed the timeout setting, and the effective timeout was so high that by the time control was returned to my sketch, it was too late to use the response body for anything constructive, causing me downstream issues that I confused with your solution not working.

I don't have a sense of how widespread missing Content-Length responses are in general, but it might be nice to have other ways to deal with it.  For example, an option to pass a (pointer to a) string of characters that if received, cause a return prior to the timeout.  In my case, the response body is fully formed XML, so it's easy to know when it's done.  Having to rely exclusively on a timeout forces one to play the game of picking a time that's long enough to most likely gather the full response, but short enough that you aren't waiting around for no reason.  But because the response time is likely to be somewhat variable, that seems like a less than ideal solution, and something more certain might be nice.

Thanks very much for your efforts.

Reply:
As mentioned below in the thread, there is a more generic way for stopping the stream of an unknown length without having to wait for the timeout to occur.